File: plugin/src/main/groovy/com/sensorsdata/analytics/android/plugin/version/SensorsDataSDKVersionHelper.java
package com.sensorsdata.analytics.android.plugin.version;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.sensorsdata.analytics.android.plugin.SensorsAnalyticsTransform;
import com.sensorsdata.analytics.android.plugin.utils.TextUtil;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class SensorsDataSDKVersionHelper {
    public static final String VERSION_KEY_SENSORDATA_VERSION_CONFIG = "SensorsDataVersionConfig";
    public static final String VERSION_KEY_CURRENT_VERSION = "SDK_VERSION";
    public static final String VERSION_KEY_DEPENDENT_SDK_VERSION = "DEPENDENT_SDK_VERSIONS";
    public static final String VERSION_KEY_PLUGIN_SDK_PATH = "com/sensorsdata/analytics/android/plugin";
    public static final String DEFAULT_MESSAGE = "当前神策 Android SDK 版本 %s 过低，请升级至 %s 及其以上版本后进行使用";

    private Map<String, List<SensorsDataSDKVersionBean>> mDependentVersionMap = new HashMap<>();
    private Map<String, String> mCurrentVersionMap = new HashMap<>();

    public SensorsDataSDKVersionHelper() {
        mCurrentVersionMap.put(VERSION_KEY_PLUGIN_SDK_PATH, SensorsAnalyticsTransform.VERSION);
    }

    /**
     * 获取版本兼容提示信息（在 SDK 获取依赖版本信息时调用）
     *
     * @param classname 当前 SDK 版本信息路径
     * @param relatedVersion 当前 SDK 依赖 SDK 的版本信息
     * @return 提示 message，如果为 "" 表示版本兼容。
     */
    public String getMessageBySDKRelyVersion(String classname, String relatedVersion) {
        if (!TextUtil.isEmpty(classname) && !TextUtil.isEmpty(relatedVersion)) {
            JsonElement jsonElement = new JsonParser().parse(relatedVersion);
            if (null != jsonElement) {
                if (jsonElement.isJsonArray()) {
                    JsonArray jsonArray = jsonElement.getAsJsonArray();
                    if (null != jsonArray && jsonArray.size() > 0) {
                        List<SensorsDataSDKVersionBean> sensorSDKBeanList = new ArrayList<>();
                        for (int i = 0; i < jsonArray.size(); i++) {
                            JsonElement jsonElementSon = jsonArray.get(i);
                            if (jsonElementSon.isJsonObject()) {
                                JsonObject jsonObjectSon = jsonElementSon.getAsJsonObject();
                                SensorsDataSDKVersionBean sensorSDKBean = SensorsDataSDKVersionBean.createSensorDataSDKBean(jsonObjectSon);
                                if (null != sensorSDKBean) {
                                    sensorSDKBeanList.add(sensorSDKBean);
                                }
                            }
                        }
                        if (sensorSDKBeanList.size() > 0) {
                            mDependentVersionMap.put(classname, sensorSDKBeanList);
                        }
                    }
                }
            }
        }
        if (null != mCurrentVersionMap && mCurrentVersionMap.size() > 0) {
            //对比版本了
            return checkSensorsSDKVersionOnAnalytic();
        }
        return "";
    }

    /**
     * 获取版本兼容提示信息（在 SDK 获取版本时调用 ）
     *
     * @param classname 当前 SDK 版本的信息的类路径，每一个 SDK 唯一，用于匹配
     * @param currentVersion 当前的版本
     * @return 提示 message，如果为 "" 表示版本兼容。
     */
    public String getMessageBySDKCurrentVersion(String classname, String currentVersion) {
        mCurrentVersionMap.put(classname, currentVersion);
        return mDependentVersionMap.size() > 0 ? checkSensorsSDKVersionOnAnalytic() : "";
    }

    /**
     * 真正检测 SDK 版本匹配
     *
     * @return 提示 message，如果为 "" 表示版本兼容。
     */
    private String checkSensorsSDKVersionOnAnalytic() {
        for (List<SensorsDataSDKVersionBean> beanList : mDependentVersionMap.values()) {
            if (beanList.size() > 0) {
                Iterator<SensorsDataSDKVersionBean> itList = beanList.iterator();
                while (itList.hasNext()) {
                    SensorsDataSDKVersionBean sensorSDKBean = itList.next();
                    if (mCurrentVersionMap.containsKey(sensorSDKBean.getSensorsDataSDKPath())) {
                        String message = sensorSDKBean.getSensorsDataSDKVersionMessage(mCurrentVersionMap.get(sensorSDKBean.getSensorsDataSDKPath()));
                        itList.remove();
                        if (!TextUtil.isEmpty(message)) {
                            return message;
                        }
                    }
                }
            }
        }
        return "";
    }
}

File: sa-gradle-plugin/src/main/kotlin/com/sensorsdata/analytics/android/plugin/version/SensorsDataSDKVersionHelper.kt
package com.sensorsdata.analytics.android.plugin.version

import com.google.gson.JsonParser
import com.sensorsdata.analytics.android.plugin.common.VersionConstant
import com.sensorsdata.analytics.android.plugin.utils.TextUtil.isEmpty

class SensorsDataSDKVersionHelper {
    private val mDependentVersionMap: MutableMap<String, ArrayList<SensorsDataSDKVersionBean>> =
        HashMap()

    private val mCurrentVersionMap: MutableMap<String, String> = HashMap()

    init {
        mCurrentVersionMap[VERSION_KEY_PLUGIN_SDK_PATH] = VersionConstant.VERSION
    }

    /**
     * 获取版本兼容提示信息（在 SDK 获取依赖版本信息时调用）
     *
     * @param classname      当前 SDK 版本信息路径
     * @param relatedVersion 当前 SDK 依赖 SDK 的版本信息
     * @return 提示 message，如果为 "" 表示版本兼容。
     */
    fun getMessageBySDKRelyVersion(classname: String, relatedVersion: String?): String {
        if (!isEmpty(classname) && !isEmpty(relatedVersion)) {
            val jsonElement = JsonParser.parseString(relatedVersion)
            if (null != jsonElement) {
                if (jsonElement.isJsonArray) {
                    val jsonArray = jsonElement.asJsonArray
                    if (null != jsonArray && jsonArray.size() > 0) {
                        val sensorSDKBeanList = ArrayList<SensorsDataSDKVersionBean>()
                        for (i in 0 until jsonArray.size()) {
                            val jsonElementSon = jsonArray[i]
                            if (jsonElementSon.isJsonObject) {
                                val jsonObjectSon = jsonElementSon.asJsonObject
                                val sensorSDKBean =
                                    SensorsDataSDKVersionBean.createSensorDataSDKBean(jsonObjectSon)
                                if (null != sensorSDKBean) {
                                    sensorSDKBeanList.add(sensorSDKBean)
                                }
                            }
                        }
                        if (sensorSDKBeanList.size > 0) {
                            mDependentVersionMap[classname] = sensorSDKBeanList
                        }
                    }
                }
            }
        }
        return if (mCurrentVersionMap.isNotEmpty()) {
            //对比版本了
            checkSensorsSDKVersionOnAnalytic()
        } else ""
    }

    /**
     * 获取版本兼容提示信息（在 SDK 获取版本时调用 ）
     *
     * @param classname      当前 SDK 版本的信息的类路径，每一个 SDK 唯一，用于匹配
     * @param currentVersion 当前的版本
     * @return 提示 message，如果为 "" 表示版本兼容。
     */
    fun getMessageBySDKCurrentVersion(classname: String, currentVersion: String): String {
        mCurrentVersionMap[classname] = currentVersion
        return if (mDependentVersionMap.isNotEmpty()) checkSensorsSDKVersionOnAnalytic() else ""
    }

    /**
     * 真正检测 SDK 版本匹配
     *
     * @return 提示 message，如果为 "" 表示版本兼容。
     */
    private fun checkSensorsSDKVersionOnAnalytic(): String {
        for (beanList in mDependentVersionMap.values) {
            if (beanList.size > 0) {
                val itList = beanList.iterator()
                while (itList.hasNext()) {
                    val sensorSDKBean = itList.next()
                    if (mCurrentVersionMap.containsKey(sensorSDKBean.sensorsDataSDKPath)) {
                        val message = sensorSDKBean.getSensorsDataSDKVersionMessage(
                            mCurrentVersionMap[sensorSDKBean.sensorsDataSDKPath]!!
                        )
                        itList.remove()
                        if (!isEmpty(message)) {
                            return message
                        }
                    }
                }
            }
        }
        return ""
    }

    companion object {
        const val VERSION_KEY_SENSORDATA_VERSION_CONFIG = "SensorsDataVersionConfig"
        const val VERSION_KEY_CURRENT_VERSION = "SDK_VERSION"
        const val VERSION_KEY_DEPENDENT_SDK_VERSION = "DEPENDENT_SDK_VERSIONS"
        const val VERSION_KEY_PLUGIN_SDK_PATH = "com/sensorsdata/analytics/android/plugin"
        const val DEFAULT_MESSAGE = "当前神策 Android SDK 版本 %s 过低，请升级至 %s 及其以上版本后进行使用"
    }
}